internal/graph/check.go

```diff
@@ -805,64 +848,181 @@ func buildTupleKeyConditionFilter(ctx context.Context, reqCtx *structpb.Struct,
+func (c *LocalChecker) checkUsersetFastPath(ctx context.Context, req *ResolveCheckRequest, iter *storage.ConditionsFilteredTupleKeyIterator) (*ResolveCheckResponse, error) {
 	ctx, span := tracer.Start(ctx, "checkUsersetFastPath")
 	defer span.End()
-
+	// Caller already verified typesys
 	typesys, _ := typesystem.TypesystemFromContext(ctx)
-	// We had just checked for the existence of the typesys in the caller.
-	// So, we are guaranteed for the presence of the context.
+	usersetDetails := buildUsersetDetails(typesys, "", tuple.GetType(req.GetTupleKey().GetUser()))
+	return c.checkMembership(ctx, req, iter, usersetDetails)
+}
 
-	reqUserType := tuple.GetType(req.GetTupleKey().GetUser())
+type usersetsChannelType struct {
+	err            error
+	objectRelation string            // e.g. group#member
+	objectIDs      storage.SortedSet // eg. [1,2,3] (no duplicates allowed, sorted)
+}
 
-	usersetsMap := make(usersetsMapType)
+// checkMembership for this model
+//
+// type user
+// type org
+//
+//	relations
+//		define viewer: [user]
+//
+// type folder
+//
+//	relations
+//		define viewer: [user]
+//
+// type doc
+//
+//	relations
+//		define viewer: viewer from parent
+//		define parent: [folder, org]
+//
+// works as follows.
+// If the request is Check(user:maria, viewer, doc:1).
+// 1. We build a map with folder#viewer:[1...N], org#viewer:[1...M] that are parents of doc:1. We send those through a channel.
+// 2. The consumer of the channel finds all the folders (and orgs) by looking at tuples of the form folder:X#viewer@user:maria (and org:Y#viewer@user:maria).
+// 3. If there is one folder or org found in step (2) that appears in the map found in step (1), it returns allowed=true immediately.
+func (c *LocalChecker) checkMembership(ctx context.Context, req *ResolveCheckRequest, iter *storage.ConditionsFilteredTupleKeyIterator, usersetDetails usersetDetailsFunc) (*ResolveCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "checkMembership")
+	defer span.End()
 
+	// since this is an unbuffered channel, producer will be blocked until consumer catches up
+	// TODO: when implementing set math operators, change to buffered. consider using the number of sets as the concurrency limit
+	usersetsChan := make(chan usersetsChannelType)
+
+	cancellableCtx, cancelFunc := context.WithCancel(ctx)
+	// sending to channel in batches up to a pre-configured value to subsequently checkMembership for.
+	pool := concurrency.NewPool(cancellableCtx, 1)
+	defer func() {
+		cancelFunc()
+		// We need to wait always to avoid a goroutine leak.
+		_ = pool.Wait()
+	}()
+	pool.Go(func(ctx context.Context) error {
+		c.produceUsersets(ctx, usersetsChan, iter, usersetDetails)
+		return nil
+	})
+
+	var finalErr error
+
+ConsumerLoop:
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case newBatch, channelOpen := <-usersetsChan:
+			if !channelOpen {
+				break ConsumerLoop
```

Comment:
instead of breaking from ConsumerLoop, would it be better to return the result (conceptually easier to understand)

Reply:
Can't return at this point since error handling etc happens after.